protocol/x/clob/e2e/reduce_only_orders_test.go

```diff
@@ -0,0 +1,392 @@
+						{
+							AssetId:  0,
+							Quantums: dtypes.NewInt(507_496_250_000),
+						},
+					},
+					PerpetualPositions: []*satypes.PerpetualPosition{
+						{
+							PerpetualId:  0,
+							Quantums:     dtypes.NewInt(99_999_850),
+							FundingIndex: dtypes.NewInt(0),
+						},
+					},
+				},
+			},
+		},
+	}
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			tApp := testapp.NewTestAppBuilder(t).WithGenesisDocFn(func() (genesis types.GenesisDoc) {
+				genesis = testapp.DefaultGenesis()
+				testapp.UpdateGenesisDocWithAppStateForModule(
+					&genesis,
+					func(genesisState *satypes.GenesisState) {
+						genesisState.Subaccounts = tc.subaccounts
+					},
+				)
+				return genesis
+			}).Build()
+			ctx := tApp.InitChain()
+
+			// Create all orders.
+			deliverTxsOverride := make([][]byte, 0)
+			deliverTxsOverride = append(
+				deliverTxsOverride,
+				constants.ValidEmptyMsgProposedOperationsTxBytes,
+			)
+
+			for _, order := range tc.ordersForFirstBlock {
+				for _, checkTx := range testapp.MustMakeCheckTxsWithClobMsg(
+					ctx,
+					tApp.App,
+					*clobtypes.NewMsgPlaceOrder(order),
+				) {
+					resp := tApp.CheckTx(checkTx)
+					require.Conditionf(t, resp.IsOK, "Expected CheckTx to succeed. Response: %+v", resp)
+
+					if order.IsStatefulOrder() {
+						deliverTxsOverride = append(deliverTxsOverride, checkTx.Tx)
+					}
+				}
+			}
+
+			// Add an empty premium vote.
+			deliverTxsOverride = append(deliverTxsOverride, constants.EmptyMsgAddPremiumVotesTxBytes)
+
+			ctx = tApp.AdvanceToBlock(2, testapp.AdvanceToBlockOptions{
+				DeliverTxsOverride: deliverTxsOverride,
+			})
+
+			// Place orders for second block
+			for _, order := range tc.ordersForSecondBlock {
+				for _, checkTx := range testapp.MustMakeCheckTxsWithClobMsg(
+					ctx,
+					tApp.App,
+					*clobtypes.NewMsgPlaceOrder(order),
+				) {
+					resp := tApp.CheckTx(checkTx)
+					require.Conditionf(t, resp.IsOK, "Expected CheckTx to succeed. Response: %+v", resp)
+
+					if order.IsStatefulOrder() {
+						deliverTxsOverride = append(deliverTxsOverride, checkTx.Tx)
+					}
+				}
+			}
+
+			// Verify expectations.
+			for orderId, exists := range tc.expectedOrderOnMemClob {
+				_, existsOnMemclob := tApp.App.ClobKeeper.MemClob.GetOrder(ctx, orderId)
+				// _, found := tApp.App.ClobKeeper.GetLongTermOrderPlacement(ctx, orderId)
+				require.Equal(t, exists, existsOnMemclob)
+			}
+
+			for orderId, expectedFillAmount := range tc.expectedOrderFillAmount {
+				exists, fillAmount, _ := tApp.App.ClobKeeper.GetOrderFillAmount(ctx, orderId)
+				require.True(t, exists)
+				require.Equal(t, expectedFillAmount, fillAmount.ToUint64())
+			}
+
+			for _, subaccount := range tc.expectedSubaccounts {
+				actualSubaccount := tApp.App.SubaccountsKeeper.GetSubaccount(ctx, *subaccount.Id)
+				require.Equal(t, subaccount, actualSubaccount)
+			}
+		})
+	}
+}
+
+func TestReduceOnlyOrderFailure(t *testing.T) {
```

Comment:
can we add conditional order tests + long-term order tests for reduce-only?

Reply:
Future PR